### PR TITLE
Added GeneratedCodeAttribute and ExcludeFromCodeCoverageAttribute to …

### LINKED ISSUE #1792: Add normal code gen'd annotations to generated code
We are generating code in the AppHost project:

https://github.com/dotnet/aspire/blob/0d1f8ea7750b69c8f221816077e9e5a5baa1bc0a/src/Aspire.Hosting/build/Aspire.Hosting.targets#L35-L44

But we aren't doing the normal "code gen'd" things to this code:

* Add a header saying it is generated code
* Add ExcludeFromCodeCoverage 
* etc

We should write this generated code like all the other generated code.

### PR DESCRIPTION
**Pull request by MatsM16**
- Fixes #1792 

Add `GeneratedCodeAttribute` and `ExcludeFromCodeCoverageAttribute` to generated project classes and properties.

Resulting project class:  
```cs
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

namespace Projects;

[global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "8.0.100-rtm.23512.16")]
[global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage(Justification = "Generated code.")]
[global::System.Diagnostics.DebuggerDisplay("Type = {GetType().Name,nq}, ProjectPath = {ProjectPath}")]
public class MyAspireProject
{
    [global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "8.0.100-rtm.23512.16")]
    public string ProjectPath { get; } = """C:\Full\path\to\MyAspireProject.csproj""";
}
```

Edit: `ExcludeFromCodeCoverageAttribute` on class makes property-attribute redundant: [ExcludeFromCodeCoverageAttribute Remarks](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.excludefromcodecoverageattribute?view=net-8.0#remarks)
Edit2: Added version to `GeneratedCodeAttribute`.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2822)

### CODE REVIEW

**Comment by MatsM16** [comment]
I would love some help on how to get the current version of the `Aspire.Hosting`-package to put in the `Version`-property on the `GeneratedCodeAttribute`.

**Comment by JamesNK** [comment]
There is also an `<auto-generated/>` comment that can be placed at the very top of the file:

```cs
// <auto-generated/>
using xxxx;
```

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 1]
```diff
@@ -36,10 +36,13 @@
         <Source>
         <![CDATA[namespace Projects%3B
 
+[global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "")]
```
Comment (eerhardt) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:39]: We should also add it on line 79 below.

**Comment by eerhardt** [comment]
> I would love some help on how to get the current version of the Aspire.Hosting-package to put in the Version-property on the GeneratedCodeAttribute.

I think there are 2 general approaches that could work:

1. We "stamp" the version number into the .targets file during the build. Similar to how we do it in 
https://github.com/dotnet/aspire/blob/3e5f36462128d19878d7e9f38b9f6e6ab9673459/src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj#L20-L21
2. We create a Task that gives us the version. But we would need to build a new Task library. See https://github.com/dotnet/aspire/issues/1825

**Comment by MatsM16** [comment]
> 1. We "stamp" the version number into the .targets file during the build. Similar to how we do it in
> ...
`MicrosoftDotnetSdkInternalPackageVersion` is available in `Aspire.Hosting.targets`.  
Currently this generates:
```cs
[global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "8.0.100-rtm.23512.16")]
```

Will `MicrosoftDotnetSdkInternalPackageVersion` always contain an acceptable version number?

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 2]
```diff
@@ -34,12 +34,24 @@
...
+//     This code was generated by a tool.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+namespace Projects%3B
+
+[global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "]]>$(MicrosoftDotnetSdkInternalPackageVersion)<![CDATA[")]
```
Comment (eerhardt) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:48]: `$(MicrosoftDotnetSdkInternalPackageVersion)` is not correct. It is not the Aspire version, it is the version of the .NET SDK we are using to build the workload.

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 3]
```diff
@@ -34,12 +34,24 @@
...
+[global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage(Justification = "Generated code.")]
 [global::System.Diagnostics.DebuggerDisplay("Type = {GetType().Name,nq}, ProjectPath = {ProjectPath}")]
 public class ]]>%(ClassName)<![CDATA[ : global::Aspire.Hosting.IProjectMetadata
 {
-  public string ProjectPath => """]]>%(ProjectPath)<![CDATA["""%3B
+    [global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "]]>$(MicrosoftDotnetSdkInternalPackageVersion)<![CDATA[")]
```
Comment (eerhardt) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:53]:
```suggestion
```

I don't think we need to add `GeneratedCode` to every member. Just marking the class as `GeneratedCode` is enough.

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 4]
```diff
@@ -34,12 +34,24 @@
...
+    public string ProjectPath { get%3B } = """]]>%(ProjectPath)<![CDATA["""%3B
```
Comment (eerhardt) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:54]: Please revert this change. There is no reason to make a field on the object just to store this constant string.

File: src/Aspire.Hosting/build/Aspire.Hosting.targets (hunk 2 above)
Comment (eerhardt) [reply src/Aspire.Hosting/build/Aspire.Hosting.targets:48]: Note also, `$(MicrosoftDotnetSdkInternalPackageVersion)` is only set during our local build. It isn't going to be set during a customer's build.

Comment (MatsM16) [reply src/Aspire.Hosting/build/Aspire.Hosting.targets]: I have added this to `Aspire.Hosting.csproj`:
```xml
<ItemGroup>
  <None Include="**/*.props;**/*.targets" ... PerformTextReplacement="True" />
</ItemGroup>
...
<ItemGroup>
  <!-- Copy the package version into Aspire.Hosting.targets,
       this is required to correctly version generated code. -->
  <TextReplacementValue Include="Version" NewValue="$(Version)" />
</ItemGroup>
```

And changed `Aspire.Hosting.targets` like this:
```xml
<AspireVersion>@Version@</AspireVersion>
<AspireVersion Condition="$(AspireVersion) == '@Version@'">$(Version)</AspireVersion>
```

Is this enough to copy the files?

I am sorry for the amount of questions and help needed, but I find the network of `targets`-, `prop-` and `csproj`-files a bit confusing.
I truly appreciate you taking the time to help me :)

Comment (eerhardt) [reply src/Aspire.Hosting/build/Aspire.Hosting.targets]: > I find the network of targets-, prop- and csproj-files a bit confusing.

I found this book to be extremely useful for learning about MSBuild. https://www.amazon.com/Inside-Microsoft-Build-Engine-Foundation/dp/0735645248. The first 3 chapters especially. https://learn.microsoft.com/en-us/visualstudio/msbuild/build-process-overview does a pretty good job too.

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 5]
```diff
@@ -34,12 +36,23 @@
...
+//     This code was generated by a tool.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+namespace Projects%3B
+
+[global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", "]]>$(AspireVersion)<![CDATA[")]
```
Comment (eerhardt) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:50]: Let's just use `null` for the version for now and log an issue especially for it. This is going to take more work than what we are doing here.

File: src/Aspire.Hosting/build/Aspire.Hosting.targets (hunk 2 above)
Comment (baronfel) [reply src/Aspire.Hosting/build/Aspire.Hosting.targets]: This book is on MS Library - I'll shoot you a link!

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 6]
```diff
@@ -34,12 +34,23 @@
     <ItemGroup>
       <AspireProjectMetadataSource Update="@(AspireProjectMetadataSource)">
         <Source>
-        <![CDATA[namespace Projects%3B
-
+        <![CDATA[//------------------------------------------------------------------------------
```
Comment (eerhardt) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:37]: In the dotnet/runtime source generators we just generate:

```C#
// <auto-generated/>
```

https://github.com/dotnet/runtime/blob/4cf19ee05d3efc5372b8f308240688bacce09208/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs#L32
https://github.com/dotnet/runtime/blob/4cf19ee05d3efc5372b8f308240688bacce09208/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs#L135

These files go into the `obj` folder, I'm not sure we need the whole comment here.

**Comment by MatsM16** [comment]
Rebase to latest `main` in hope of fixing helix progresql tests
```
C:\Users\cloudtest\.nuget\packages\microsoft.dotnet.helix.sdk\8.0.0-beta.24161.7\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(89,5): error : Work item Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests in job ffb3d4d1-39ab-45bb-870d-954d4bf753c9 has failed. [D:\a\_work\1\s\tests\send-to-helix.proj]
C:\Users\cloudtest\.nuget\packages\microsoft.dotnet.helix.sdk\8.0.0-beta.24161.7\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(89,5): error : Failure log: https://helix.dot.net/api/2019-06-17/jobs/ffb3d4d1-39ab-45bb-870d-954d4bf753c9/workitems/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests/console [D:\a\_work\1\s\tests\send-to-helix.proj]
```

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution!
